Issue #1005 Add PUT operation and the codec for the same

albinsuresh opened the pull request:
A few assumptions made during the operation definition:
- The design was limited to simple get and put operations only.
- A sequence of PUT operations get resolved to the last PUT operation. A need for a different type to represent the resolved value does not exist at the moment as there is not metadata associated with operations. Hence PUT operation can be reused. A new type can be added if the need arises in future.
- Since the initial iteration is not gonna have expiry support the metadata has been completely omitted. Addition of metadata for expiry calculations will depend upon the decision taken on #1015.

ljacomet commented on clustered/client/src/main/java/org/ehcache/clustered/client/internal/store/operations/BaseOperation.java:
```diff
@@ -0,0 +1,65 @@
+ */
+public abstract class BaseOperation<K> {
+
+  public enum Code {
```
I fail to see the point of this enum. It is awkward in the sense that being declared in the `BaseOperation` it will have to be close to all sub classes, thus braking encapsulation. And even for serialisation, it does not really help more than simply having directly a byte field inside the operation.

albinsuresh replied:
I agree. All I needed was a simple byte in the `BaseOperation` class to represent the type of operation. But the reason I chose the enum was to clearly map the operation type with a value. If no such mapping exists anywhere, there will be if conditions/switch blocks in the decode methods of codecs that checks for different values before decoding. Code like this:
```java
byte code = buffer.get();
switch(code) {
  case 0:
    //do stuff for PUT
  case 1:
    // do stuff for another operation
  default:
    // something else
}
```

So the mapping will end up defined in the conditional blocks in the decode methods. That is what I was trying to avoid. Define the mapping somewhere centrally. And use it while decoding. And enum provided the added benefit of readability at places where the decoding is done.

So that was the rationale behind the addition of enums. Does it make sense?

lorban replied:
What I think Louis has in mind is to augment the `OperationCodecFactory` class to contain a pair of maps: one `Map<Class<? extends BaseOperation>,? extends OperationCodec>` for encoding where you can lookup the codec using the operation and another one `Map<Byte, ? extends OperationCodec>` for decoding where you can lookup the codec using the op byte.

That way, you'd get rid of the need to use switch blocks as that would centralize that logic in a single location.

ljacomet replied:
Or if you believe the enum is the right way to go, push it all the way. It becomes the central place where operation characteristics, codecs and others are recored and the rest of the code becomes more generic.

ljacomet commented on clustered/client/src/main/java/org/ehcache/clustered/client/internal/store/operations/codecs/PutOperationCodec.java:
```diff
@@ -0,0 +1,86 @@
+      throw new AssertionError(this.getClass().getName() + " can not decode operation of type " + opCode);
+    }
+
+    byte[] keyBlob = new byte[buffer.getInt()];
```
Consider slicing the ByteBuffer instead of allocating another byte[], a bit more math to compute bounds, but less allocations in the end.

ljacomet commented on clustered/client/src/main/java/org/ehcache/clustered/client/internal/store/operations/codecs/OperationsCodec.java:
```diff
@@ -0,0 +1,58 @@
+
+  public ByteBuffer encode(BaseOperation<K> operation) {
+    BaseOperation.Code opCode = operation.getOpCode();
+    switch (opCode) {
```
Having `switch` here feels like encapsulation breakage. Not that I have a specific recommendation just now.

ljacomet replied:
Scratch that - `OperationCodecFactory` could be based on the operation concrete type, using a `Map<Class<? extends BaseOperation>,? extends OperationCodec>`

albinsuresh commented:
Addressed some of the review feedback:
- Improved the decoding logic by avoiding additional allocations
- Did some refactoring to improve the abstractions

lorban commented on clustered/client/src/main/java/org/ehcache/clustered/client/internal/store/operations/codecs/OperationCodecFactory.java:
```diff
@@ -0,0 +1,57 @@
+    return operationCodeVsCodecMap.get(code);
+  }
+
+  private OperationCodec<K> getNewCodec(OperationCode code) {
```
This looks rather convoluted still. You could replace that switch with a `getCodec()` method in the `OperationCode` enum.

albinsuresh replied:
Yeah I had though about it. That'd mean that I'll have to move the codec initialization to the enum. The codecs would require the key and value serializers for initialization. So that'd mean that I'll have to pass both the serializers at the time of enum constant initialization. I shouldn't be doing that, right?

lorban replied:
Good point, let's not do that and leave that the way it is - unless someone else has a better suggestion.

ljacomet replied:
This could still be improved by having the enum return an `OperationCodecFactory` which would return the proper `OperationCodec` through a method taking the different `Serializer`s.

albinsuresh replied:
Well things could get a bit weird if I embark on that path. Or may be I'm not understanding your suggestion properly: 

If I add a method at the enum level returning an `AbstractOperationCodecFactory` which has a structure something like this: 

```java
abstract class AbstractOperationCodecFactory<K, V> {
  public abstract OperationCodec<K> getOperationCodec(Serializer<K> keySerializer, Serializer<V> valueSerializer);
}
```
and a concrete implementation, like for example `PutOperationCodecFactory` with a structure like this:

```java
public class PutOperationCodecFactory<K, V> extends AbstractOperationCodecFactory<K, V> {

  OperationCodec<K> getOperationCodec(Serializer<K> keySerializer, Serializer<V> valueSerializer) {
    return new PutOperationCodec<K, V>(keySerializer, valueSerializer);
  }
}
```
If we take this path, then for every encode method call, we'll end up calling:

Enum.getOperationCodec(keySer, valueSer).encode(operation)

Every _getOperationCodec_ call is gonna return a new instance of the codec with the provided serializer pairs and finally we end up with too many `OperationCodec` instances. So I'm not sure if this is the right thing to do.

Now if we decide to cache the instance of codec on the first _getOperationCodec_ call and then decide to reuse it on subsequent calls( singleton like approach), we'll end up with an enum that'll be tied to a single pair of serializers which in-turn ties the enum to a single cache. I don't think that's what we want either.

These were the two approaches that came to my mind. Did you have something totally different in mind? May be I'm missing something.

ljacomet replied:
I am not sure I follow some of your logic and maybe there is something I am missing.
In the context of this method `private OperationCodec<K> getNewCodec(OperationCode code)` instead of the switch, you do:
```java
return code.getOperationCodecFactory().newCodec(keySerializer, valueSerializer);
```
And that is it, the rest of this class remains the same, you just encapsulated the codec selection inside the enum without changing any of the lifecycle existing as of now.

However, it is true I did not do a complete lifecycle analysis and maybe there is something not right there.

albinsuresh replied:
"In the context of the _getNewCodec_ method". Okay. Now I get it. Have updated the PR.

albinsuresh commented:
In the last iteration I've done some nasty renaming which has resulted in a slightly convoluted diff view:
The old `OperationCodecFactory` has been renamed as `OperationCodecProvider`. And the name `OperationCodecFactory` was reused as the name of the new abstract base class that is extended by all operation specific codec factories like `PutOperationCodecFactory`. So just watch out for these.

lorban commented on clustered/client/src/main/java/org/ehcache/clustered/client/internal/store/operations/codecs/PutOperationCodec.java:
```diff
@@ -0,0 +1,96 @@
+
+  public ByteBuffer encode(BaseOperation<K> operation) {
+    if (!(operation instanceof PutOperation)) {
+      throw new AssertionError(this.getClass().getName() + " can only decode " + PutOperation.class.getName());
```
I would rather throw `IllegalArgumentException` here.

lorban commented on clustered/client/src/main/java/org/ehcache/clustered/client/internal/store/operations/codecs/PutOperationCodec.java:
```diff
@@ -0,0 +1,96 @@
+
+    OperationCode opCode = OperationCode.valueOf(buffer.get());
+    if (opCode != PUT) {
+      throw new AssertionError(this.getClass().getName() + " can not decode operation of type " + opCode);
```
As above, I would rather throw `IllegalArgumentException` here.

lorban commented on clustered/client/src/test/java/org/ehcache/clustered/client/internal/store/operations/codecs/PutOperationCodecTest.java:
```diff
@@ -0,0 +1,76 @@
+    assertEquals(key, putOperation.getKey());
+    assertEquals(value, putOperation.getValue());
+  }
+}
```
I would add some more tests:
 - 2 tests that check that the code throws when a wrong operation is being encoded / decoded
 - a test that decodes what's been encoded then asserting that you get the original key/value pair

lorban commented:
Please rebase, then it's good to go.
